### ISSUE TITLE
Integrate scala-java8-compat in the standard library ?

### ISSUE THREAD

**Issue by smarter** [scala/scala-dev#556]
I don't really get why https://github.com/scala/scala-java8-compat is still a separate library now that Scala is Java 8+ only. As more and more Java libraries (including the JDK itself) start taking advantages of Java 8 Optional/streams/..., using them from Scala becomes increasingly more cumbersome.
A built-in converter between Option and Optional would be especially useful: it's not a big implementation cost, and it means that people won't reimplement it over and over and over again: https://github.com/sbt/util/blob/develop/internal/util-logging/src/main/scala/sbt/util/InterfaceUtil.scalahttps://github.com/scalacenter/bloop/blob/master/frontend/src/main/scala/bloop/util/JavaCompat.scala https://github.com/superpixelhq/jdot/blob/master/src/main/scala/com/superpixel/jdot/util/ScalaConverters.scala https://github.com/proglang/jgs/blob/master/GradualConstraints/Tests/src/test/java/de/unifreiburg/cs/proglang/jgs/util/Interop.scala https://github.com/thaytharma/dc18-webauthnapplication/blob/master/webauthn-server-core/src/main/scala/com/yubico/scala/util/JavaConverters.scala https://github.com/beatmadsen/xcs-main/blob/master/src/main/scala/com/madsen/util/JavaConversions.scala https://github.com/atomist-attic/artifact-source/blob/master/src/main/scala/com/atomist/util/Utils.scala

**Comment by NthPortal** [comment]
I don't know of the migration cost, but I think it would make sense to migrate them from `scala.compat.java8` to `scala.compat.java` (and `scala.concurrent.java8` to `scala.concurrent.compat.java` or maybe just `scala.compat.java` as well? I'm not sure), since there's no reason for the package to be exclusive to Java 8 things.

**Comment by martijnhoekstra** [comment]
From the opposite perspective, should narrow(er) integration with OpenJDK be in the stdlib at all (which means required re-implementation of those OpenJDK parts that it has close interop with for scala-js and scala-native), or should stuff like JavaConversions be moved *out* of the stdlib to compat libraries to reduce pressure on scala-js and scala-native to implement interfaces that only exist to be hidden by a compat package?

**Comment by sjrd** [comment]
Indeed. Please let's try not to put *more* stuff related only to interop with Java in the stdlib. We should have less of those, not more.

**Comment by hrhino** [comment]
Agreed on both `scala.compat.java` and on moving more compat stuff in.

Most of the Scala programmers I know handle Java APIs with great care as it is, and adding another library (especially if it's as stable as I hope it can be) is a negligible cost compared to the cost of referencing more JRE classes in the stdlib.

**Comment by SethTisue** [comment]
there ought to be a "Java interop guide" on https://docs.scala-lang.org. I've just made a ticket on that at https://github.com/scala/docs.scala-lang/issues/1146 . among other things, it could call people's attention to scala-java8-compat's existence

**Comment by smarter** [comment]
A concrete example where the lack of modern Java interop in the standard library is hurting us: JDK 11 adds `String#lines`, this takes precedence over Scala's `StringLike#lines`, the former returns a `Stream` thus breaking everyone's codebase. If we had something built-in to handle streams this wouldn't be too bad, but I doubt every library is going to start depending on `scala-java8-compat` just for this. (A better solution would be to use`StringLike#linesIterator`, unfortunately it's deprecated so not an option if you like `-Xfatal-warnings`).

**Comment by dwijnand** [comment]
JFYI that example is tracked and in discussion at https://github.com/scala/bug/issues/11125.

**Comment by szeiger** [comment]
Another option is the reintroduction of an aggregated standard library like the now defunct `scala-library-all`. `scala-java8-compat` could become a separate library module in the standard scala repository and be part of a new `scala-library-all` which would be the recommended library on the JVM.

@lrytz suggests a `scala-library-core` without excessive JVM compatibility stuff (like `JavaConverters`) and having `scala-library` be the aggregated version. I'd prefer this approach, too, but I'm worried about tooling compatibility (e.g. sbt expecting an actual `scala-library` JAR).

**Comment by SethTisue** [comment]
discussed at team meeting just now. tentative consensus:

* reintegrate: yes
* scala-library-all: no
* scala-library-core: not explicitly discussed, but probably not?

key arguments:

* discoverability of external libraries remains poor (Scala Platform hasn't happened, easier pulling of additional JARs into the REPL hasn't happened)
* `JavaConverters` is already in and the distinction between Java 8+ and Java 8- no longer applies
* scala-java8-compat is tightly integrated with collections internals, is isn't in fact independent code

hopefully the reintegration can be done in a manner that only causes minor inconvenience for e.g. Scala.js

**Comment by SethTisue** [comment]
@lrytz will assess the stuff that *isn't* about collections
